internal/app/admin/http/handler.go

```diff
@@ -102,30 +110,108 @@ func cacheDumpHandler(o *orchestrator.Orchestrator) http.HandlerFunc {
 // the map of requests is converted to a slice of just the keys,
 // since the bool value is meaningless.
-// TODO(lisalu): More intelligent unmarshalling of DiscoveryResponse.
 func resourceToString(resource cache.Resource) (string, error) {
 	var requests []*v2.DiscoveryRequest
 	for request := range resource.Requests {
 		requests = append(requests, request)
 	}
 
 	resourceString := &marshallableResource{
-		Resp:           resource.Resp,
+		Resp:           marshalDiscoveryResponse(resource.Resp),
 		Requests:       requests,
 		ExpirationTime: resource.ExpirationTime,
 	}
 
 	return stringify.InterfaceToString(resourceString)
 }
 
+type marshalledDiscoveryResponse struct {
+	VersionInfo  string
+	Resources    *xDSResources
+	Canary       bool
+	TypeURL      string
+	Nonce        string
+	ControlPlane *envoy_api_v2_core.ControlPlane
+}
+
+type xDSResources struct {
+	Endpoints []*v2.ClusterLoadAssignment
+	Clusters  []*v2.Cluster
+	Routes    []*v2.RouteConfiguration
+	Listeners []*v2.Listener
+	Secrets   []*envoy_api_v2_auth.Secret
+	Runtimes  []*envoy_service_discovery_v2.Runtime
+}
+
+func marshalDiscoveryResponse(resp *v2.DiscoveryResponse) *marshalledDiscoveryResponse {
+	if resp != nil {
+		marshalledResp := marshalledDiscoveryResponse{
+			VersionInfo:  resp.VersionInfo,
+			Canary:       resp.Canary,
+			TypeURL:      resp.TypeUrl,
+			Resources:    marshalResources(resp.Resources),
+			Nonce:        resp.Nonce,
+			ControlPlane: resp.ControlPlane,
+		}
+		return &marshalledResp
+	}
+	return nil
+}
+
+func marshalResources(Resources []*any.Any) *xDSResources {
+	var marshalledResources xDSResources
+	for _, resource := range Resources {
+		switch resource.TypeUrl {
+		case resource2.EndpointType:
+			e := &v2.ClusterLoadAssignment{}
+			err := ptypes.UnmarshalAny(resource, e)
+			if err == nil {
+				marshalledResources.Endpoints = append(marshalledResources.Endpoints, e)
+			}
+		case resource2.ClusterType:
+			c := &v2.Cluster{}
+			err := ptypes.UnmarshalAny(resource, c)
+			if err == nil {
+				marshalledResources.Clusters = append(marshalledResources.Clusters, c)
+			}
+		case resource2.RouteType:
+			r := &v2.RouteConfiguration{}
+			err := ptypes.UnmarshalAny(resource, r)
+			if err == nil {
+				marshalledResources.Routes = append(marshalledResources.Routes, r)
+			}
+		case resource2.ListenerType:
+			l := &v2.Listener{}
+			err := ptypes.UnmarshalAny(resource, l)
+			if err == nil {
+				marshalledResources.Listeners = append(marshalledResources.Listeners, l)
+			}
+		case resource2.SecretType:
+			s := &envoy_api_v2_auth.Secret{}
+			err := ptypes.UnmarshalAny(resource, s)
+			if err == nil {
+				marshalledResources.Secrets = append(marshalledResources.Secrets, s)
+			}
+		case resource2.RuntimeType:
+			r := &envoy_service_discovery_v2.Runtime{}
+			err := ptypes.UnmarshalAny(resource, r)
+			if err == nil {
+				marshalledResources.Runtimes = append(marshalledResources.Runtimes, r)
+			}
+		default:
```

Comment:
Do you think we should add a catch all `[]*any.Any` in the `xDSResources` struct in this scenario and include that in the output?

Reply:
Added